go/ql/lib/semmle/go/frameworks/Fasthttp.qll

```diff
@@ -0,0 +1,515 @@
+            ]) and
+          this = f.getACall()
+        )
+      }
+
+      override DataFlow::Node getAPathArgument() { result = this.getArgument(1) }
+    }
+
+    /**
+     * A function that can be used as a sanitizer for XSS
+     */
+    class HtmlQuoteSanitizer extends SharedXss::Sanitizer {
+      HtmlQuoteSanitizer() {
+        exists(DataFlow::CallNode c |
+          c.getTarget()
+              .hasQualifiedName(packagePath(),
+                ["AppendHTMLEscape", "AppendHTMLEscapeBytes", "AppendQuotedArg"])
+        |
+          this = c.getArgument(1)
+        )
+      }
+    }
+
+    /**
+     * A function that sends HTTP requests
+     * Get* send a HTTP GET request.
+     * Post send a HTTP POST request.
+     * these Functions first arguments is a URL.
+     */
+    class RequestForgerySink extends RequestForgery::Sink {
+      RequestForgerySink() {
+        exists(Function f |
+          f.hasQualifiedName(packagePath(), ["Get", "GetDeadline", "GetTimeout", "Post"]) and
+          this = f.getACall().getArgument(1)
+        )
+      }
+
+      override DataFlow::Node getARequest() { result = this }
+
+      override string getKind() { result = "URL" }
+    }
+
+    /**
+     * A function that sends HTTP requests.
+     * First argument of following functions need Additional steps.
+     * look at URI module, additional steps part for more information.
+     */
+    class RequestForgerySinkDo extends RequestForgery::Sink {
+      RequestForgerySinkDo() {
+        exists(Function f |
+          f.hasQualifiedName(packagePath(), ["Do", "DoDeadline", "DoTimeout", "DoRedirects"]) and
+          this = f.getACall().getArgument(0)
+        )
+      }
+
+      override DataFlow::Node getARequest() { result = this }
+
+      override string getKind() { result = "URL" }
+    }
+
+    /**
+     * A function that create initial connection to a TCP address.
+     * Following Functions only accept TCP address + Port in their first argument
+     */
+    class RequestForgerySinkDial extends RequestForgery::Sink {
+      RequestForgerySinkDial() {
+        exists(Function f |
+          f.hasQualifiedName(packagePath(),
+            ["DialDualStack", "Dial", "DialTimeout", "DialDualStackTimeout"]) and
+          this = f.getACall().getArgument(0)
+        )
+      }
+
+      override DataFlow::Node getARequest() { result = this }
+
+      override string getKind() { result = "TCP Addr + Port" }
+    }
+  }
+
+  /**
+   * Provide modeling for fasthttp.URI Type
+   */
+  module URI {
+    /**
+     * The additioanl steps that can be used in fasthttp framework.
+     * Fasthttp has its own uri creating/manipulation methods and these methods usually are used in code.
+     * Pred can be an user controlled value like any potential part of URL and succ is the URI instance.
+     * So if we called a method like `URIInstance.SetHost(pred)` then the URIInstance is succ.
+     */
+    class UriAdditionalStep extends AdditionalStep {
```

Comment:
Okay, I've had a closer look at this and I have thought of a better way of doing this. I assume that we want any kind of taint to propagate along these steps. Then we can model them using models-as-data (MaD) in a yml file, like we do [here](https://github.com/github/codeql/blob/0ad338f04a6cb32e00df5461a090647bda9d243c/go/ql/lib/ext/github.com.astaxie.beego.model.yml). There is some documentation of the format [here](https://github.com/github/codeql/blob/0ad338f04a6cb32e00df5461a090647bda9d243c/go/ql/lib/semmle/go/dataflow/ExternalFlow.qll#L1-L74). (This wasn't possible for Afero because MaD can only model function calls, and for Afero we wanted a step across a struct literal.)

To get you started, I think one line would look like this:
```
["github.com/valyala/fasthttp", "URI", False, "SetHost", "", "", "Argument[0]", "Argument[-1]", "taint", "manual"]
```

Reply:
The way you test these steps would have to change a little. Maybe you could do an InlineFlowTest - search for that string to see some examples.